src/engine.cc

```diff
@@ -173,6 +173,27 @@ void EngineController::SetPosition(const std::string& fen,
   search_.reset();
 }
 
+std::string EngineController::GetCurrentPositionFen() {
+  if (tree_ != nullptr) {
+    return tree_->HeadPosition().GetFen();
+  } else if (current_position_) {
```

Comment:
I think it would be cleaner to split this function into two. How about that:

1. Make `current_position_` non `std::optional` (initialize it to `ChessBoard::kStartposFen, {}` on construction in in new_game`.
2. Remove `else` branch in `if (current_position_)` in `EngineController::Go`.
3. Have function `Position FoldPosition(const CurrentPosition& pos)` which would do what this branch does (no need to put it in a class). (or some better name than `FoldPosition` but it's good to have a name that implies that it's not a O(1) operation... `BuildPosition()`? `ApplyPositionMoves()`?)
4. Call `FoldPosition(current_position_).GetFen()` instead of `GetCurrentPositionFen()`.
5. (optional) rename `CurrentPosition` (struct name, not the variable) to something better. :)

Reply:
Note that now the position is never taken from the tree - this is implied in the above and makes in practice, as a `position` followed by `fen` would give the fen for the tree head with the previous version instead of the just set position.